### PR TITLE
Document WebAssembly target feature expectations

### LINKED ISSUE #109807: Can not disable sign-ext feature for wasm32 target
I tried to build a wasm binary with MVP opcode only, so I disabled the features:

```bash
RUSTFLAGS='-C link-arg=-zstack-size=32768 -C target-feature=-mutable-globals,-sign-ext,-multivalue,-simd128'
 cargo build --release --target=wasm32-unknown-unknown
```

I expected to see no [sign-extension opcode](https://github.com/WebAssembly/sign-extension-ops/blob/master/proposals/sign-extension-ops/Overview.md) in the wasm binary.

Instead, this happened: 
```bash
$ wasm-tools parse -t target/wasm32-unknown-unknown/release/output.wasm -o hello.wat
$ cat hello.wat | grep i32.ext
                            i32.extend8_s
```

### Meta

`rustc --version --verbose`:
```
rustc 1.70.0-nightly (ec2f40c6b 2023-03-30)
binary: rustc
commit-hash: ec2f40c6b04f0e9850dd1f454e8639d319f4ed9b
commit-date: 2023-03-30
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.0
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

### LINKED ISSUE #119811: WebAssembly platform requirements are not well documented
### Location

README.md and WebAssembly book (https://rustwasm.github.io/docs/book/)

### Summary

In [version 1.70](https://github.com/rust-lang/rust/blob/master/RELEASES.md#compatibility-notes-5) Rust added some requirements for WebAssembly targets. This makes Wasm binaries compiled with Rust (at least with default options/out of the box std) incompatible with older runtime environments, in particular not-so-old iOS safari (https://webassembly.org/features/). Here is for example the support for Safari iOS 15.5 simulator:

safari-15_5-wasm-support

I searched but did not find any place where these requirements are written down. I think that this is a problem.

### LINKED ISSUE #128475: Generated WebAssembly unexpectedly requires reference types
### Code

This can be replicated with a minimal `test.rs`:

```rs
fn main() {}
```

Compile it to WASM with `rustc test.rs -o test.wasm --target wasm32-unknown-unknown`

Then, check if the generated WASM uses reference types:

```
$ wasm-validate --disable-reference-types test.wasm
0000260: error: call_indirect reserved value must be 0
```

### Version it worked on

It most recently worked on: `nightly-2024-07-19` (and previous stable rust versions)

On those versions, compiling the above `test.rs` and then running `wasm-validate` with `--disable-reference-types` doesn't fail.

### Version with regression

`rustc --version --verbose`:
```
rustc 1.82.0-nightly (28a58f2fa 2024-07-31)
binary: rustc
commit-hash: 28a58f2fa7f0c46b8fab8237c02471a915924fe5
commit-date: 2024-07-31
host: x86_64-unknown-linux-gnu
release: 1.82.0-nightly
LLVM version: 19.1.0
```

@rustbot modify labels: +regression-from-stable-to-nightly -regression-untriaged

This seems to have been caused by https://github.com/rust-lang/rust/pull/127513

The compatibility note there says:

> The WebAssembly target features multivalue and reference-types are now both enabled by default, but generated code is not affected by default

However, the generated code seems to be indeed affected by default. Interestingly, reference types seem to be used unnecessarily, as using `wasm2wat` to turn the generated `test.wasm` into wat and then `wat2wasm` to turn it back into wasm removes the reference types.

As-is, this breaks compatibility with the WebAssembly 1.0 spec.

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit is a result of the discussion on #128475 and incorporates parts of #109807 as well. This is all done as a new page of documentation for the `wasm32-unknown-unknown` target which previously did not exist. This new page goes into details about the preexisting target and additionally documents the expectations for WebAssembly features and code generation.

The tl;dr is that LLVM will enable features over time after most engines have had support for awhile. Compiling without features requires `-Ctarget-cpu=mvp` to rustc plus `-Zbuild-std` to Cargo.

Closes #109807
Closes #119811
Closes #128475

EDIT: for the `relnotes` label please see [this comment](https://github.com/rust-lang/rust/pull/128511#issuecomment-2299051004) for recommended release notes.

### CODE REVIEW

**Comment by alexcrichton** [comment]
If this is accepted then I will additionally comment on https://github.com/rust-lang/rust/pull/127513 with a recommended rewording of the release notes to indicate that this does affect WebAssembly binaries.

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (daxpedda) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:9]:
```suggestion
JavaScript environments as there is no standard for what functions can be
```

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (daxpedda) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:89]:
```suggestion
This target can be cross-compiled from any host.
```

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (daxpedda) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:95]:
```suggestion
`println!` doesn't work in the standard library. It's recommended to test the
```

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (daxpedda) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:121]:
```suggestion
they're enabled in LLVM by default. There is currently no exact number of
```

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (daxpedda) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:151]: Should we also link to the [Rust Reference](https://doc.rust-lang.org/reference/attributes/codegen.html#wasm32-or-wasm64)?

**Review by daxpedda — APPROVED** [review]

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (bjorn3) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:169]: *sufficient

File: src/doc/rustc/src/platform-support/wasm32-wasip1.md
Comment (bjorn3) [inline src/doc/rustc/src/platform-support/wasm32-wasip1.md:139]: The wasm32-wasip1-threads target also enables bulk-memory and atomics, right?

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (kleisauke) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:144]:
```suggestion
$ export RUSTFLAGS=-Ctarget-cpu=mvp
```

File: src/doc/rustc/src/platform-support/wasm32-wasip1.md
Comment (alexcrichton) [reply src/doc/rustc/src/platform-support/wasm32-wasip1.md]: Indeed!

File: src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md
Comment (bb010g) [inline src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md:169]:
```suggestion
[`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md) but two more features
```

File: src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md
Comment (bb010g) [inline src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md:176]:
```suggestion
[`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md), but note that the
```

File: src/doc/rustc/src/platform-support/wasm32-wasip1.md
Comment (bb010g) [inline src/doc/rustc/src/platform-support/wasm32-wasip1.md:139]:
```suggestion
same as [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md). See the
```

File: src/doc/rustc/src/platform-support/wasm32-wasip2.md
Comment (bb010g) [inline src/doc/rustc/src/platform-support/wasm32-wasip2.md:68]:
```suggestion
same as [`wasm32-unknown-unknown`](./wasm32-unknown-unknown.md). See the
```

File: src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md
Comment (bb010g) [inline src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md:170]: "similar to `wasm32-unknown-unknown` but two more features are included" doesn't specify whether features originally included in `wasm32-unknown-unknown` are no longer included.

**Review by bb010g — COMMENTED** [review]
A wording nit & spelling nits.

File: src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md
Comment (stephanemagnenat) [reply src/doc/rustc/src/platform-support/wasm32-wasip1-threads.md:170]: Note that there is a typo in the second `unknokwn` at line 169.

**Comment by workingjubilee** [comment]
@alexcrichton Is it possible to get an rmake test for the workarounds here, or landing a test in rust-lang/cargo?

**Comment by alexcrichton** [comment]
I was looking into that yeah and was hoping I could combine `-Clto` somehow to get a test in-tree, but the conclusion ended up being that it didn't have the right behavior of rewriting target features per function that I thought was happening. That means that such a test requires `-Zbuild-std` which while there is [one test in tree that does this](https://github.com/rust-lang/rust/blob/fac77538022b32b4298833b541feb457ca4548b0/tests/run-make/compiler-builtins/rmake.rs#L42) that test looks relatively complicated and specialized so I was wary to try to copy it.

I can look into adding a test in Cargo, yeah, but it feels like an odd location for such a test given that the regressions, if any, would likely come from this repo instead of Cargo. One option would be to refactor the existing Zbuild-std test in this repo, but I'm also not sure if folks are keen on expanding that due to the amount of time such a test would take compiling the standard library.

**Comment by jieyouxu** [comment]
> I was looking into that yeah and was hoping I could combine `-Clto` somehow to get a test in-tree, but the conclusion ended up being that it didn't have the right behavior of rewriting target features per function that I thought was happening. That means that such a test requires `-Zbuild-std` which while there is [one test in tree that does this](https://github.com/rust-lang/rust/blob/fac77538022b32b4298833b541feb457ca4548b0/tests/run-make/compiler-builtins/rmake.rs#L42) that test looks relatively complicated and specialized so I was wary to try to copy it.

Yeah, unfortunately the compiler-builtins test takes a really long time to build the stdlib again, since it relies on very specific configurations. I don't have a good solution for this either, if it requires building yet another customized version of std.

**Comment by alexcrichton** [comment]
ping @GuillaumeGomez would you be able to review this? Or should I try to get a different sign-off?

**Comment by GuillaumeGomez** [comment]
I'm often assigned for such PRs but I have nothing to do with this part. Please pick someone else. ^^'

**Comment by alexcrichton** [comment]
@workingjubilee would you perhaps be able to lend an r+?

**Comment by jieyouxu** [comment]
I'll take a look. r? jieyouxu

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (jieyouxu) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:96]: Remark (for T-compiler): this may or may not be significant for T-compiler as per [Tier 2 target policy](https://doc.rust-lang.org/rustc/target-tier-policy.html#tier-2-target-policy):

> The approving teams may additionally require that a subset of tests pass in CI, such as enough to build a functional "hello world" program, ./x.py test --no-run, or equivalent "smoke tests".

File: src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md
Comment (jieyouxu) [inline src/doc/rustc/src/platform-support/wasm32-unknown-unknown.md:146]: Note (for T-compiler): this may be significant since if a common engine does not support a feature in LLVM's enabled-by-default feature set then the user has to rebuild std with engine-matching flags.

**Review by jieyouxu — APPROVED** [review]
The docs look good to me! However, as this is modifying (or rather, making explicit / documented) the baseline expectations of Tier 2 targets `wasm32-unknown-unknown`, `wasm32-wasip1-threads`, `wasm32-wasip1` and `wasm32-wasip2`, I think this will need T-compiler signoff.

**Comment by jieyouxu** [comment]
Nominating this PR for T-compiler sign-off and pinging T-release for compatibility amendments.

@rustbot label +I-compiler-nominated

### For T-compiler

cc @rust-lang/compiler as this might require a T-compiler FCP.

This PR adds target documentation for the `wasm32-unknown-unknown` Tier 2 target. The PR also explicitly documents/clarifies the baseline expectations for the `wasm32-unknown-unknown`, `wasm32-wasip1-threads`, `wasm32-wasip1` and `wasm32-wasip2` Tier 2 targets with respect to enabled-by-default [WebAssembly Proposals](https://github.com/WebAssembly/proposals) (at the time of writing and review):

- `wasm32-unknown-unknown`: `multivalue`, `mutable-globals`, `reference-types`, `sign-ext`
- `wasm32-wasip1-threads`: `wasm32-unknown-unknown` enabled-by-default proposals + `bulk-memory`, `atomics`
- `wasm32-wasip1`, `wasm32-wasip2`: `wasm32-unknown-unknown` enabled-by-default proposals

Furthermore, this PR explains in the target documentation for the wasm32 targets how to target *Minimal Viable Product (MVP)* wasm, with none of the above proposals enabled.

This PR was opened in response in part due to the [LLVM 19 update](https://github.com/rust-lang/rust/pull/127513) which caused more wasm proposals to be enabled-by-default, namely `reference-types` and `multivalue`, and this was noticed by a user in https://github.com/rust-lang/rust/issues/128475. It is expected that more wasm proposals will become enabled-by-default over time.

In particular, this means that the compatibility notes in the LLVM update PR are not accurate:

> The WebAssembly target features `multivalue` and `reference-types` are now both enabled by default, but generated code is not affected by default.

Since we found out in https://github.com/rust-lang/rust/issues/128475 that the generated code is, in fact, affected by `multivalue` and `reference-types` proposals.

### For T-release

@rustbot label +relnotes

cc @rust-lang/release 

As we've discovered in https://github.com/rust-lang/rust/issues/128475, the compatibility notes for https://github.com/rust-lang/rust/pull/127513 are not accurate, as the `multivalue` and `reference-types` now enabled-by-default wasm proposals *do* in fact affect generated code. The target docs introduced by this PR explains how to generate code targeting MVP wasm which do not have these wasm proposals enabled-by-default.

**Comment by jieyouxu** [comment]
@alexcrichton can you double-check if my summaries above are accurate?

**Comment by alexcrichton** [comment]
That all looks accurate to me, thanks for writing that up! Also I apologize I thought I had commented with updated release notes for https://github.com/rust-lang/rust/pull/127513 but apparently I never actually did that. I would recommend the following for updated release notes (commenting here to reduce locations where folks are pinged, but I can copy this over into the LLVM 19 update too:

> The WebAssembly target features `multivalue` and `reference-types` are now both enabled by default. These two features both have subtle changes implied for generated WebAssembly binaries. For the `multivalue` feature, WebAssembly target support has changed when upgrading to LLVM 19. Support for generating functions with multiple returns no longer works and `-Ctarget-feature=+multivalue` has a different meaning than it did in LLVM 18 and prior. There is no longer any supported means to generate a module that has a function with multiple returns in WebAssembly from Rust source code. For the `reference-types` feature the encoding of immediates in the `call_indirect`, a commonly used instruction by the WebAssembly backend, has changed. Validators and parsers which don't understand the `reference-types` proposal will no longer accept modules produced by LLVM due to this change in encoding of immediates. Additionally these features being enabled are encoded in the `target_features` custom section and may affect downstream tooling such as `wasm-opt` consuming the module. Generating a WebAssembly module that disables default features requires `-Zbuild-std` support from Cargo and more information can be found at [rust-lang/rust#128511](https://github.com/rust-lang/rust/pull/128511).

Otherwise I can also try to add some color to this as well. This change is a reflection of the evolution of the WebAssembly standard itself. WebAssembly is being extended beyond its initial capabilities and as changes are proposed and implemented they eventually make their way back to toolchains. The maintainers of LLVM and various other languages are aware of the consequences of enabling new proposals by default and it's not something that's done willy nilly. This matches what the Emscripten toolchain is doing and proposals are only enabled once a sufficient number of runtimes have supported proposals for a sufficient period of time (neither of which is precisely defined, however).

**Comment by apiraino** [comment]
Discussed in T-compiler triage meeting [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-08-22/near/464392557).

To expand on the `T-release` label addition, we think that:
> we should just make sure to communicate this to users. At the very least, there should be release notes and a mention in the release blog. We may also want a dedicated blog post earlier than that.

@rustbot label -I-compiler-nominated

**Review by jieyouxu — APPROVED** [review]
Since we double-checked that this does not need a T-compiler FCP, this LGTM now.

**Comment by danielhjacobs** [comment]
Is the minimum support Safari version for this target version 14: https://caniuse.com/wasm-reference-types?

**Comment by danielhjacobs** [comment]
I really just wish Rust hadn't enabled extra target features by default when it's so easy to enable extra target features and so hard to disable them. Why can't that be down to the crates to enable?

**Comment by daxpedda** [comment]
Just for clarification: Rust didn't enable those features by default, LLVM did.
But please see [this blog post](https://blog.rust-lang.org/2024/09/24/webassembly-targets-change-in-default-target-features.html) for full context.

**Comment by workingjubilee** [comment]
Please open a new issue or a discussion on Zulip.

**Pull request merged**
